common/src/main/java/org/opensearch/ml/common/model/LocalRegexGuardrail.java

```diff
@@ -0,0 +1,265 @@
+        }
+        return true;
+    }
+
+    public Boolean validateStopWordsSingleIndex(String input, String indexName, List<String> fieldNames) {
+        SearchRequest searchRequest;
+        AtomicBoolean hitStopWords = new AtomicBoolean(false);
+        String queryBody;
+        Map<String, String> documentMap = new HashMap<>();
+        for (String field : fieldNames) {
+            documentMap.put(field, input);
+        }
+        Map<String, Object> queryBodyMap = Map
+                .of("query", Map.of("percolate", Map.of("field", "query", "document", documentMap)));
+        CountDownLatch latch = new CountDownLatch(1);
+        ThreadContext.StoredContext context = null;
+
+        try {
+            queryBody = AccessController.doPrivileged((PrivilegedExceptionAction<String>) () -> gson.toJson(queryBodyMap));
+            SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+            XContentParser queryParser = XContentType.JSON.xContent().createParser(xContentRegistry, LoggingDeprecationHandler.INSTANCE, queryBody);
+            searchSourceBuilder.parseXContent(queryParser);
+            searchSourceBuilder.size(1); //Only need 1 doc returned, if hit.
+            searchRequest = new SearchRequest().source(searchSourceBuilder).indices(indexName);
+            if (isStopWordsSystemIndex(indexName)) {
+                context = client.threadPool().getThreadContext().stashContext();
+                ThreadContext.StoredContext finalContext = context;
+                client.search(searchRequest, ActionListener.runBefore(new LatchedActionListener(ActionListener.<SearchResponse>wrap(r -> {
+                    if (r == null || r.getHits() == null || r.getHits().getTotalHits() == null || r.getHits().getTotalHits().value == 0) {
+                        hitStopWords.set(true);
+                    }
+                }, e -> {
+                    log.error("Failed to search stop words index {}", indexName, e);
+                    hitStopWords.set(true);
+                }), latch), () -> finalContext.restore()));
+            } else {
+                client.search(searchRequest, new LatchedActionListener(ActionListener.<SearchResponse>wrap(r -> {
+                    if (r == null || r.getHits() == null || r.getHits().getTotalHits() == null || r.getHits().getTotalHits().value == 0) {
+                        hitStopWords.set(true);
+                    }
+                }, e -> {
+                    log.error("Failed to search stop words index {}", indexName, e);
+                    hitStopWords.set(true);
+                }), latch));
+            }
+        } catch (Exception e) {
+            log.error("[validateStopWords] Searching stop words index failed.", e);
+            latch.countDown();
+            hitStopWords.set(true);
+        } finally {
+            if (context != null) {
+                context.close();
+            }
+        }
+
+        try {
+            latch.await(5, SECONDS);
```

Comment:
Suggest change to non-blocking way with action listener. Which thread pool this part will block?

Reply:
Will address it in next pr.